optaplanner-core/src/main/java/org/optaplanner/core/impl/partitionedsearch/DefaultPartitionedSearchPhase.java

```diff
@@ -143,14 +148,43 @@ public void solve(DefaultSolverScope<Solution_> solverScope) {
                 stepEnded(stepScope);
                 phaseScope.setLastCompletedStepScope(stepScope);
             }
-        } finally {
+        } catch (RuntimeException e) {
             // If a partition thread throws an Exception, it is propagated here
             // but the other partition thread won't finish any time soon, so we need to terminate them
             childThreadPlumbingTermination.terminateChildren();
+        } finally {
+            List<Runnable> partitionSolvers = threadPoolExecutor.shutdownNow();
```

Comment:
This should actually handle the thread clean-up you mentioned. This is even covered by the new test.

Sorry for the messy code below but that's the cost of being super-defensive.